# Initial time series forecasting: S&P 500

- simple one step ahead forecasting using a sliding window approach
- tested linear reg, naive, svm and nn models. 
- No hpyerparameter tuning for svm or nn yet. 

In [1]:
# import some shit
%matplotlib widget
# %matplotlib notebook 

import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

# model evalution metrics
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

# data preprocessing
from sklearn.preprocessing import normalize

# %matplotlib inline
# import some data
sp_500 = pd.read_csv('../test_data/GSPC.csv')
sp_500

Date         Open         High          Low        Close  \
0      1950-01-03    16.660000    16.660000    16.660000    16.660000   
1      1950-01-04    16.850000    16.850000    16.850000    16.850000   
2      1950-01-05    16.930000    16.930000    16.930000    16.930000   
3      1950-01-06    16.980000    16.980000    16.980000    16.980000   
4      1950-01-09    17.080000    17.080000    17.080000    17.080000   
...           ...          ...          ...          ...          ...   
17213  2018-05-31  2720.979980  2722.500000  2700.679932  2705.270020   
17214  2018-06-01  2718.699951  2736.929932  2718.699951  2734.620117   
17215  2018-06-04  2741.669922  2749.159912  2740.540039  2746.870117   
17216  2018-06-05  2748.459961  2752.610107  2739.510010  2748.800049   
17217  2018-06-06  2753.250000  2772.389893  2748.459961  2772.350098   

         Adj Close      Volume  
0        16.660000     1260000  
1        16.850000     1890000  
2        16.930000     2550000  
3        16.980000     2010000  
4        17.080000     2520000  
...            ...         ...  
17213  2705.270020  4235370000  
17214  2734.620117  3684130000  
17215  2746.870117  3376510000  
17216  2748.800049  3517790000  
17217  2772.350098  3651640000  

[17218 rows x 7 columns]

In [2]:
# plot the data
sp_500.plot(x='Date',figsize=(10,10),subplots=True)
plt.tight_layout()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

# Calculate returns on open price as data for prediction tasks

In [3]:
# calculate returns on 
returns = sp_500[['Date','Open']]
returns['Shift'] = sp_500['Open'].shift(periods=-1) # use pandas shift method to create shift daily open price one time period forward
returns['returns'] = returns['Shift'] - returns['Open'] # calculates the 24hr return
returns.drop(labels=returns.index[-1],axis=0,inplace=True) # need to delete last row as the shift value is nan and therefore no return could be calculated
display(returns)

fig, ax = plt.subplots(figsize=(12,5))
returns.plot(x='Date',y='returns',ax=ax)

<ipython-input-3-51bb1f54035d>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  returns['Shift'] = sp_500['Open'].shift(periods=-1) # use pandas shift method to create shift daily open price one time period forward
<ipython-input-3-51bb1f54035d>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  returns['returns'] = returns['Shift'] - returns['Open'] # calculates the 24hr return
C:\Users\tristan\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a co

Date         Open        Shift    returns
0      1950-01-03    16.660000    16.850000   0.190000
1      1950-01-04    16.850000    16.930000   0.080000
2      1950-01-05    16.930000    16.980000   0.050000
3      1950-01-06    16.980000    17.080000   0.100000
4      1950-01-09    17.080000    17.030001  -0.049999
...           ...          ...          ...        ...
17212  2018-05-30  2702.429932  2720.979980  18.550048
17213  2018-05-31  2720.979980  2718.699951  -2.280029
17214  2018-06-01  2718.699951  2741.669922  22.969971
17215  2018-06-04  2741.669922  2748.459961   6.790039
17216  2018-06-05  2748.459961  2753.250000   4.790039

[17217 rows x 4 columns]

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

<AxesSubplot:xlabel='Date'>

# 0. Useful functions

In [4]:
# write function to perform sliding window over data: transform time series in supervised learning problem

def slide_window(data, slide_step_size): # column of data, integer
    # initialize input array
    num_rows = len(data) - slide_step_size
    array = np.zeros((num_rows, slide_step_size + 1))
    
    # loop through data and populate array
    for i in range(num_rows):
        # input features
        array[i,0:slide_step_size+1] = data[i:i+slide_step_size+1]
        # target feature
        array[i,-1] = data[i+slide_step_size]
        # show pattern
        print(array[i,0:slide_step_size],' : ',array[i,slide_step_size])
    return array 

dummy_data = [1,2,3,4,5,6,7,8,9,10]
dummy_result = slide_window(dummy_data,5)  

[1. 2. 3. 4. 5.]  :  6.0
[2. 3. 4. 5. 6.]  :  7.0
[3. 4. 5. 6. 7.]  :  8.0
[4. 5. 6. 7. 8.]  :  9.0
[5. 6. 7. 8. 9.]  :  10.0


# 1. Data preporation: 

Re-framing time series data as supervised machine learning problem using a sliding window approach.

In [5]:
# inputs to select and prepare data
column = 'returns' # which stock price feature to use for prediction
window_length = 5 # how many previous time steps is used to set up supvervised learning problem

In [6]:
# testing training data split
# training_data = slide_window(list(sp_500[column][-2000:-500]),window_length)
# print('\n*************************************************************************************************\n')
# test_data = slide_window(list(sp_500[column][-500:]),window_length)

# rather use returns data
training_data = slide_window(list(returns[column][-2000:-500]),window_length)
print('\n*************************************************************************************************\n')
test_data = slide_window(list(returns[column][-500:]),window_length)

# fourier filtered results
inverse_transform_filtered_real = np.real(inverse_transform_filtered) # only take real part, this comes from taking the inverse fourier transform after filtering out low psd frequencies

training_data = slide_window(inverse_transform_filtered_real[-2000:-500],window_length)
print('\n*************************************************************************************************\n')
test_data = slide_window(inverse_transform_filtered_real[-500:],window_length)

15.839844   9.830078   4.369873]  :  8.130126999999902
[16.680176 15.839844  9.830078  4.369873  8.130127]  :  4.959961000000021
[15.839844  9.830078  4.369873  8.130127  4.959961]  :  -2.8300779999999577
[ 9.830078  4.369873  8.130127  4.959961 -2.830078]  :  3.9599609999995664
[ 4.369873  8.130127  4.959961 -2.830078  3.959961]  :  0.020019000000502274
[ 8.130127  4.959961 -2.830078  3.959961  0.020019]  :  15.630126999999902
[ 4.959961 -2.830078  3.959961  0.020019 15.630127]  :  6.5500489999999445
[-2.830078  3.959961  0.020019 15.630127  6.550049]  :  -5.910155999999915
[ 3.959961  0.020019 15.630127  6.550049 -5.910156]  :  0.11010699999997087
[ 0.020019 15.630127  6.550049 -5.910156  0.110107]  :  2.239990999999918
[15.630127  6.550049 -5.910156  0.110107  2.239991]  :  2.119873000000098
[ 6.550049 -5.910156  0.110107  2.239991  2.119873]  :  -10.51001000000042
[ -5.910156   0.110107   2.239991   2.119873 -10.51001 ]  :  8.270019000000502
[  0.110107   2.239991   2.119873 -10.51

NameError: name 'inverse_transform_filtered' is not defined

In [10]:
training_data.shape

(1495, 6)

In [11]:
test_data.shape

(495, 6)

In [12]:
# normalize training data
training_data = normalize(training_data)
test_data = normalize(test_data)

# 2. Visualize data

In [16]:
# plot training and testing data
fig, ax = plt.subplots(figsize=(10,5))

returns[column][-2000:-500].plot(ax=ax,style='k-',label='Training data') # replace returns with sp_500 for other data plotting
returns[column][-500:].plot(ax=ax,style='r-',label='test data')
plt.plot(returns[column][-500+window_length:].index,test_data[:,-1],'o',label='test data') # important to match time by start 5 (length of time window) after where segmented our testing and training data
plt.legend(loc=0)

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

# 3. Predictive modelling

Train three models:
- baseline model: naive model which says tomorrow's price is today's price.
- multivariate regression
- support vector regression

## 3.1 Linear regression

In [14]:
# Train simple linear regression model and lets see what we get
from sklearn.linear_model import LinearRegression

# data
X = training_data[:,0:-1]
Y = training_data[:,-1]

# model
reg_model = LinearRegression().fit(X,Y)

# view trained model parameters
reg_model.coef_

array([-0.03072718, -0.01011928, -0.00605004, -0.00044199,  0.00534345])

#### 3.1.1 Look at linear regression coefficients

In [15]:
for i in range(len(reg_model.coef_)):
    print(reg_model.coef_[i])

-0.030727182279870823
-0.010119281124213748
-0.006050037794133537
-0.00044198622327739625
0.005343450862550363


For non-stationary data, eg only using Open price:
- The weights of the previous time step is extremely strong / large, relative to previous time steps / features. ie this model is saying the best predictor for tomorrow is today. Essentially, predict tomorrow's price as todays price.

For stationary data, eg using returns:
- all weights are more similar in magnitude. This means that the model is not simply relying on X_t-1 to predict X_t

#### 3.1.2 Test model on test data, ie last 500 days

In [16]:
df_test_data = pd.DataFrame(data=test_data)
df_test_data

0         1         2         3         4         5
0   -0.622359 -0.527365  0.033185 -0.392553  0.413502 -0.091498
1   -0.619227  0.038966 -0.460931  0.485530 -0.107436  0.394082
2    0.048274 -0.571044  0.601519 -0.133102  0.488226  0.231672
3   -0.564952  0.595103 -0.131682  0.483018  0.229201  0.153304
4    0.598987 -0.132541  0.486171  0.230697  0.154304  0.557005
..        ...       ...       ...       ...       ...       ...
490 -0.603725  0.420325 -0.181906 -0.458244 -0.066424  0.459732
491  0.525937 -0.227612 -0.573384 -0.083114  0.575246 -0.070705
492 -0.205166 -0.516839 -0.074917  0.518518 -0.063732  0.642066
493 -0.518415 -0.075146  0.520099 -0.063927  0.644023  0.190376
494 -0.086812  0.600846 -0.073851  0.744010  0.219933  0.155152

[495 rows x 6 columns]

In [17]:
# use trained model to predict

# loop through each test data pattern and predict result
predicted_results = []
print('Prediction\tReal values')
for i in range(500-window_length):
    X_test = test_data[i,0:-1]
    prediction = reg_model.predict(X_test.reshape(1,-1))
    predicted_results.append(prediction)
    print(prediction,'\t',test_data[i,-1])
    
# full prediction
predictions = reg_model.predict(test_data[:,0:-1])

Prediction	Real values
[0.05707665] 	 -0.09149809672693349
[0.05106724] 	 0.394082447697819
[0.03375813] 	 0.23167229986380616
[0.04357979] 	 0.15330374683710682
[0.01115173] 	 0.5570050540706848
[0.03061204] 	 -0.9769017637438805
[0.02080881] 	 -0.3174382738965262
[0.02647792] 	 0.42161626053212625
[0.03524803] 	 0.3364512105630386
[0.03776405] 	 0.27767494205415566
[0.05586257] 	 -0.04578461747127184
[0.03337854] 	 -0.17562731553815908
[0.00207986] 	 0.2784254820029834
[0.00642726] 	 0.14488985826206457
[0.01434829] 	 0.5955278648134456
[0.03862205] 	 0.2342717500765962
[0.03476591] 	 0.3985425690266487
[0.01206969] 	 0.11782910532150741
[0.01507137] 	 0.23043196778410835
[0.00169481] 	 -0.09990436786532204
[0.00744873] 	 0.09393920442024244
[0.00034497] 	 0.13522076809027897
[0.01587795] 	 0.5894991917272764
[0.01718537] 	 -0.5117597368143834
[0.03248854] 	 0.5756005215535962
[0.0245625] 	 -0.36233813614832494
[0.0204695] 	 0.06466550428406521
[0.01682944] 	 -0.2821618829141122
[0.0

### 3.1.3 Model evaluation

In [18]:
# use sklearn metric methods to calc rmse and mae
mse = mean_squared_error(test_data[:,-1],predictions)
mae = mean_absolute_error(test_data[:,-1],predictions)

print('RMSE: ',np.sqrt(mse))
print('MAE: ',mae)

RMSE:  0.40624498456287134
MAE:  0.33153581191446285


Results without filtering:
- RMSE:  0.40624498456287134
- MAE:  0.33153581191446285

Results with filtering:
- RMSE:  0.3814258615492118
- MAE:  0.3079928467008209

## 3.2 Naive model

In [19]:
#  a naive time series predictive model predicts the next results as the current ie x[i+1] = x[i]
# ie this model will return n-1 values for n time stamps
def naive_model(data):
    preds = np.zeros(len(data))
    preds[1:] = data[0:-1]
    return  preds

### 3.2.1 predict and evaluate

In [20]:
# call naive model function
naive_predictions = naive_model(test_data[:,-1])

# evaluate predictions
mse = mean_squared_error(test_data[1:,-1],naive_predictions[1:])
mae = mean_absolute_error(test_data[1:,-1],naive_predictions[1:])

print('RMSE: ',np.sqrt(mse))
print('MAE: ',mae)

RMSE:  0.5726196993518156
MAE:  0.45291626640029387


## 3.3 Support vector machine

In [21]:
from sklearn.svm import LinearSVR

# train model
svm_regres = LinearSVR(max_iter=1000).fit(training_data[:,0:-1],training_data[:,-1])

# predict
svm_predictions = svm_regres.predict(test_data[:,0:-1])

# evaluate
mse = mean_squared_error(test_data[:,-1],svm_predictions[:])
mae = mean_absolute_error(test_data[:,-1],svm_predictions[:])

print('RMSE: ',np.sqrt(mse))
print('MAE: ',mae)

RMSE:  0.40567125484409194
MAE:  0.32881194023358284


Results without filtering:
- RMSE:  0.4054163757214758
- MAE:  0.3283511990590936

Results with filtering:
- RMSE:  0.3816363957338525
- MAE:  0.30785723389532904

## 3.4 Multilayer perceptron neural network

In [22]:
from sklearn.neural_network import MLPRegressor

# train neural network
nn_regres = MLPRegressor(hidden_layer_sizes=(100,100,100),shuffle=False,random_state=1, 
                         max_iter=1000,verbose=1).fit(training_data[:,0:-1],training_data[:,-1])

# make predictions
nn_predictions = nn_regres.predict(test_data[:,0:-1])

# evaluate
mse = mean_squared_error(test_data[:,-1],nn_predictions[:])
mae = mean_absolute_error(test_data[:,-1],nn_predictions[:])

print('RMSE: ',np.sqrt(mse))
print('MAE: ',mae)

Iteration 1, loss = 0.08466199
Iteration 2, loss = 0.08236010
Iteration 3, loss = 0.08220483
Iteration 4, loss = 0.08148397
Iteration 5, loss = 0.08131939
Iteration 6, loss = 0.08084859
Iteration 7, loss = 0.08064404
Iteration 8, loss = 0.08028393
Iteration 9, loss = 0.08000622
Iteration 10, loss = 0.07968571
Iteration 11, loss = 0.07932335
Iteration 12, loss = 0.07898744
Iteration 13, loss = 0.07862230
Iteration 14, loss = 0.07820382
Iteration 15, loss = 0.07785314
Iteration 16, loss = 0.07742386
Iteration 17, loss = 0.07703966
Iteration 18, loss = 0.07663369
Iteration 19, loss = 0.07620716
Iteration 20, loss = 0.07577037
Iteration 21, loss = 0.07531422
Iteration 22, loss = 0.07490435
Iteration 23, loss = 0.07447419
Iteration 24, loss = 0.07402689
Iteration 25, loss = 0.07355347
Iteration 26, loss = 0.07308077
Iteration 27, loss = 0.07262085
Iteration 28, loss = 0.07212841
Iteration 29, loss = 0.07163103
Iteration 30, loss = 0.07120767
Iteration 31, loss = 0.07073315
Iteration 32, los

Results without filtering:
- RMSE:  0.5356554382964639
- MAE:  0.4010986581137097

Results with filtering:
- RMSE:  0.47539037753112584
- MAE:  0.35201690512468486

# 4. Plot results of different models

In [23]:
# this function computes the error between two vectors / arrays, give the one is a vector of predicted values and the other is of real values
def error(real_data,predicted_data):
    error = np.zeros(len(real_data))
    error = (real_data - predicted_data) / real_data
    return error

In [24]:
# plot prediction against actual + training data
fig, ax = plt.subplots(2,1,figsize=(9,10),sharex=True)

# test and real y values data
# returns[column][-500:].plot(ax=ax[0],style='o-',linewidth=3,label='real values',markersize=5) # for plotting unormalized values
ax[0].plot(returns[column][-500+window_length:].index,test_data[:,-1],'o-',linewidth=3,label='real values',markersize=5) # plotting normalized training data
# predict y values
ax[0].plot(returns[column][-500+window_length:].index,predicted_results[:],'o-',label='linear regression prediction',markersize=5)
ax[0].plot(returns[column][-500+window_length+1:].index,naive_predictions[1:],'.--',label='naive prediction',markersize=5)
ax[0].plot(returns[column][-500+window_length:].index,svm_predictions[:],'.--',label='svm prediction',markersize=5)
ax[0].plot(returns[column][-500+window_length:].index,nn_predictions[:],'.--',label='nn prediction',markersize=5)

ax[0].legend()
ax[0].set_title('Real values vs model predictions')

# plot error plot
error_linreg = error(np.array(test_data[:,-1]),predictions)
error_naive = error(np.array(test_data[:,-1]),naive_predictions)
error_svm = error(np.array(test_data[:,-1]),svm_predictions)
error_nn = error(np.array(test_data[:,-1]),nn_predictions)

ax[1].plot(returns[column][-500+window_length:].index,error_linreg,'r-',label='linear reg error')
ax[1].plot(returns[column][-500+window_length+1:].index,error_naive[1:],'-',label='naive error')
ax[1].plot(returns[column][-500+window_length:].index,error_svm[:],'-',label='svm error')
ax[1].plot(returns[column][-500+window_length:].index,error_nn[:],'-',label='nn error')
ax[1].set_title('Error signal for predictive models')
ax[1].set_xlabel('Days since 1950 for s&p500')
ax[1].legend()
ax[1].set_ylim([-10,10])

# titles and save figures
title_string = 'S&P500 predictions _ y is '+str(column)+'_ window len is '+ str(window_length)
fig.suptitle(title_string)
plt.tight_layout()
fig_name = '../results/univariate_single_step_ahead/'+title_string+'.png'
plt.savefig(fig_name,facecolor='w')

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

<ipython-input-23-786cd53917c3>:4: RuntimeWarning: divide by zero encountered in true_divide
  error = (real_data - predicted_data) / real_data


## 4.1 Discussion of results

- from the graph above we can see that our suspiciouns have been confirmed: the linear regression is essentially saying the next value is the previous. This can be seen as the predicted results is almost just the actual results shift forward.
- another validation of the above hypothesis is the fact that the naive model performs near identically to linear regression
- if you zoom in you can see all models seem to favour yesterday's result as the best for today. 
- as you expand and contract the slide window, the linear reg and svm models perform very similary. The neural network however smoothens the results a lot more the longer the sliding window.

# Denoising data using FFT

In [8]:
# import scipy fft functions
from scipy.fft import fft, ifft, fftfreq

In [9]:
# dataframe to np array
length_of_time = -2000
signal = np.array(returns['returns'][length_of_time:])
signal

array([  2.400024,  -6.400024, -30.539917, ...,  22.969971,   6.790039,
         4.790039])

In [10]:
# plot original signal
fig,ax = plt.subplots(figsize=(12,5))
ax.plot(returns['returns'][length_of_time:].index,signal,'-')

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [11]:
# apply discrete fourier transform
fft_coefficients = fft(signal)
fft_coefficients

array([1678.150024    -0.j        ,  145.60226937 +12.50087687j,
        183.89946634+709.8176786j , ..., -105.58107383 +34.61480548j,
        183.89946634-709.8176786j ,  145.60226937 -12.50087687j])

In [12]:
# plot amplitude vs frequency 
n = len(signal)

# get frequencies and psd
freqs = fftfreq(signal.shape[0])
psd = np.abs(fft_coefficients)/n # psd is amplitude/N

# plot psd
fig,ax = plt.subplots()
ax.plot(freqs[0:int(n/2)],psd[0:int(n/2)])
ax.set_ylabel('Power spectrum')
ax.set_xlabel('Frequencies')

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

Text(0.5, 0, 'Frequencies')

- Volume data: we can see there is a low frequency component that is very large and many low amplitude frequency components.
- Returns data: looks very much like a random process

In [13]:
sp_500['Volume'][length_of_time:]

15218    3896410000
15219    6136700000
15220    5067080000
15221    6435770000
15222    3968500000
            ...    
17213    4235370000
17214    3684130000
17215    3376510000
17216    3517790000
17217    3651640000
Name: Volume, Length: 2000, dtype: int64

In [14]:
# plot inverse fourier transform as sanity check
inverse_fft = ifft(fft_coefficients)
fig,ax = plt.subplots(figsize=(10,5))
ax.plot(returns['returns'][length_of_time:].index,inverse_fft,'-',label='Inverse fourier')
ax.plot(returns['returns'][length_of_time:].index,returns['returns'][length_of_time:],'.',label='Real data')
ax.legend()
plt.tight_layout()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

C:\Users\tristan\anaconda3\lib\site-packages\numpy\core\_asarray.py:83: ComplexWarning: Casting complex values to real discards the imaginary part
  return array(a, dtype, copy=False, order=order)


## Denoise by removing frequencies with low amplitude

In [15]:
# try denoise data
psd_indices = psd > 0.6 # mask
fft_filtered = fft_coefficients*psd_indices

inverse_transform_filtered = ifft(fft_filtered)

# plot this
fig,ax = plt.subplots(figsize=(12,5))
ax.plot(returns['returns'][length_of_time:].index,returns['returns'][length_of_time:],'-',label='Real data')
ax.plot(returns['returns'][length_of_time:].index,inverse_transform_filtered,'-',label='Inverse fourier filtered')
ax.legend()
plt.tight_layout()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

C:\Users\tristan\anaconda3\lib\site-packages\numpy\core\_asarray.py:83: ComplexWarning: Casting complex values to real discards the imaginary part
  return array(a, dtype, copy=False, order=order)


"The Fourier transform is only able to char-
acterize truly periodic and stationary signals, as time is stripped out via the
integration in (2.18a). For a signal with non-stationary frequency content, such
as a musical composition, it is important to simultaneously characterize the
frequency content and its evolution in time." - Steve Burton text book

In [33]:
# see which frequencies are left over
blah = freqs*psd_indices

for i in range(len(blah)):
    if np.abs(blah[i]) > 0:
        print(blah[i])

0.0275
0.052000000000000005
0.07200000000000001
0.0935
0.10350000000000001
0.1075
0.11800000000000001
0.121
0.132
0.146
0.1555
0.1695
0.1715
0.17550000000000002
0.176
0.1845
0.1875
0.1925
0.2085
0.2565
0.275
0.28250000000000003
0.2915
0.2985
0.3005
0.303
0.309
0.314
0.3325
0.3355
0.3375
0.34500000000000003
0.363
0.3795
0.386
0.4045
0.41600000000000004
0.4225
0.445
-0.445
-0.4225
-0.41600000000000004
-0.4045
-0.386
-0.3795
-0.363
-0.34500000000000003
-0.3375
-0.3355
-0.3325
-0.314
-0.309
-0.303
-0.3005
-0.2985
-0.2915
-0.28250000000000003
-0.275
-0.2565
-0.2085
-0.1925
-0.1875
-0.1845
-0.176
-0.17550000000000002
-0.1715
-0.1695
-0.1555
-0.146
-0.132
-0.121
-0.11800000000000001
-0.1075
-0.10350000000000001
-0.0935
-0.07200000000000001
-0.052000000000000005
-0.0275


# Spectogram

In [8]:
from scipy import signal

f, t, Sxx = signal.spectrogram(sp_500['Volume'][-2000:], fs=1,noverlap=49,nperseg=50)

fig,ax = plt.subplots()
ax.pcolormesh(t, f, Sxx, shading='gouraud')
plt.ylabel('Frequency [Hz]')
plt.xlabel('Time [days]')
plt.ylim([0,0.1])
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [35]:
Sxx

array([[9.58479078e+16, 5.79929610e+16, 4.40271729e+16, 3.04900550e+17,
        8.05027473e+17, 8.41540301e+16, 2.34070069e+17, 6.38414276e+16,
        2.28790476e+17],
       [1.51919949e+19, 4.87945679e+18, 1.25542374e+19, 1.17805820e+19,
        1.49233646e+19, 2.09295413e+18, 1.46712384e+17, 4.02015160e+18,
        6.66351448e+18],
       [5.64977205e+18, 2.59407831e+17, 2.55876946e+18, 1.22144225e+19,
        6.36846299e+18, 9.11685989e+18, 9.41295411e+17, 2.61620804e+18,
        2.23486859e+18],
       [9.25290296e+18, 1.15628828e+18, 9.54101968e+16, 4.23414979e+18,
        1.65064824e+18, 2.03933291e+18, 6.32992504e+18, 7.17245650e+17,
        6.60948320e+18],
       [2.49118849e+18, 1.05113388e+18, 5.68430180e+17, 1.34876478e+18,
        4.87163606e+18, 5.12846521e+18, 8.55174775e+17, 1.61474581e+17,
        3.42385870e+18],
       [6.55044708e+18, 2.88896699e+17, 2.25633082e+18, 5.46578912e+17,
        1.34240021e+18, 1.01170596e+18, 1.86492905e+17, 1.20205146e+15,
        2.5

# Multi resolution analysis: wavelet transfroms

In [3]:
import pywt

In [29]:
data = sp_500['Volume'][-2000:]/1e9
index = sp_500['Volume'][-2000:].index

# Create wavelet object and define parameters
w = pywt.Wavelet('sym5')
maxlev = pywt.dwt_max_level(len(data), w.dec_len) # max level of decomposition ie scale, is determined by length of signal and discrete wavelet
# maxlev = 2 # Override if desired
print("maximum level is " + str(maxlev))
threshold = 0.25 # Threshold for filtering

# Decompose into wavelet components, to the level selected:
coeffs = pywt.wavedec(data, 'sym5', level=maxlev)

#cA = pywt.threshold(cA, threshold*max(cA))
plt.figure(figsize=(10,15))
for i in range(1, len(coeffs)):
    plt.subplot(maxlev, 1, i)
    plt.plot(coeffs[i],label='Wavelet convolution result')
    coeffs[i] = pywt.threshold(coeffs[i], threshold*max(coeffs[i]))
    plt.plot(coeffs[i],label='Wavelet convolution after threshold applied')
    plt.legend()
    plt.title('Scale '+str(len(coeffs)-i))

# inverse transfer: filtered wavelet decomposition back to time-domain signal 
datarec = pywt.waverec(coeffs, 'sym5')

plt.figure()
plt.subplot(2, 1, 1)
plt.plot(index[:], data[:])
plt.title("Raw signal")
plt.subplot(2, 1, 2)
plt.plot(index[:], datarec[:])
plt.title("De-noised signal using wavelet techniques")

plt.tight_layout()
plt.show()

maximum level is 7


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [24]:
coeffs[6]

array([ 0.85932648, -0.75737687,  0.76355347, -0.34913312, -0.16279403,
       -0.76114776,  0.        ,  0.        ,  0.        ,  0.        ,
       -0.        ,  0.        ,  0.12235676,  0.26248507, -0.06482176,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
       -0.        , -0.63698799,  0.        ,  0.        ,  0.38769524,
       -0.56960617,  0.        , -0.10405331,  0.        ,  0.07282566,
        0.89455122,  0.16268321, -0.        ,  0.42022552, -0.32984069,
       -0.13124772,  0.212767  , -0.        ,  0.        , -0.        ,
        0.        ,  0.        , -0.04108364,  0.43650019, -0.25840387,
        1.35908357,  0.78448203,  0.        ,  0.72527156, -0.        ,
        0.        , -0.        ,  0.        ,  0.        ,  0.        ,
       -0.46010958, -0.        , -0.        , -0.        ,  0.        ,
        0.        ,  0.        , -0.        , -0.        ,  0.        ,
       -0.        , -0.37279977, -0.        , -0.        ,  0.  